state-chain/pallets/cf-environment/src/lib.rs

```diff
@@ -497,46 +506,53 @@ impl<T: Config> Pallet<T> {
 		let min_fee_required_per_tx = bitcoin_fee_info.min_fee_required_per_tx();
 		let fee_per_output_utxo = bitcoin_fee_info.fee_per_output_utxo();
 
-		let calculate_utxos_and_change = |spendable_utxos: Vec<_>| {
-			let total_fee = spendable_utxos
-				.iter()
-				.map(|utxo| bitcoin_fee_info.fee_for_utxo(utxo))
-				.sum::<u64>() + fee_per_output_utxo +
-				min_fee_required_per_tx;
-
-			spendable_utxos
-				.iter()
-				.map(|utxo| utxo.amount)
-				.sum::<u64>()
-				.checked_sub(total_fee)
-				.map(|change_amount| (spendable_utxos, change_amount))
-		};
-
 		match utxo_selection_type {
-			UtxoSelectionType::SelectAllForRotation => {
-				let spendable_utxos: Vec<_> = BitcoinAvailableUtxos::<T>::take()
-					.into_iter()
-					.filter(|utxo| utxo.net_value(&bitcoin_fee_info) > 0u64)
-					.collect();
-
-				if spendable_utxos.is_empty() {
-					return None
-				}
-				calculate_utxos_and_change(spendable_utxos)
-			},
 			UtxoSelectionType::SelectForConsolidation =>
 				BitcoinAvailableUtxos::<T>::try_mutate(|available_utxos| {
-					let utxos_to_consolidate = select_utxos_for_consolidation(
-						available_utxos,
-						&bitcoin_fee_info,
-						Self::consolidation_parameters(),
-					);
-
-					if utxos_to_consolidate.is_empty() {
-						Err(())
+					let mut old_utxos = if let Some(cf_traits::EpochKey {
+						key: cf_chains::btc::AggKey { previous: Some(prev_key), current },
+						..
+					}) = T::BitcoinKeyProvider::active_epoch_key()
+					{
+						let (old_utxos, stale): (Vec<Utxo>, Vec<Utxo>) = available_utxos
+							.extract_if(|utxo| utxo.deposit_address.pubkey_x != current)
+							.partition(|utxo| utxo.deposit_address.pubkey_x == prev_key);
+
+						// Stale utxos are removed from storage and discarded
+						if !stale.is_empty() {
+							Self::deposit_event(Event::<T>::StaleUtxosDiscarded { utxos: stale });
+						}
+						old_utxos
 					} else {
-						calculate_utxos_and_change(utxos_to_consolidate).ok_or(())
+						Default::default()
+					};
+
+					// Only send up to certain limit number of utxo at a time.
+					// The rest are returned to storage after consolidation.
+					let consolidation_parameter = Self::consolidation_parameters();
+					let mut to_send = old_utxos
+						.drain(
+							..sp_std::cmp::min(
+								old_utxos.len(),
+								consolidation_parameter.consolidation_size as usize,
+							),
+						)
+						.collect::<Vec<_>>();
+
+					// If consolidation is required, add into the same batch
+					if available_utxos.len() > consolidation_parameter.consolidation_size as usize {
```

Comment:
above, we limit the number of utxos for selection. But now here we're potentially adding more to that set, so we can go over the limit. We could add a test for this case too. To make sure that that the number of utxos selected doesn't exceed the consolidation size. We can also add a debug_assert after this `if` that the `to_send.len() < consolidation_size`

Reply:
So the logic in this case could be:
1. Rotate
2. Send to new vault
3. Full, wait until next block
4. Consolidate

which is fine. The primary thing is that we don't try to sign too many payloads at once

Reply:
also not really sure what the intention of this `if` statement is? 
It's currently saying "if we have more utxos associated with the current key than maximum number we are allowed to consolidate then run this code" but that doesn't seem relevant? If anything at all, it should be the `params.consolidation_threshold`? In which case this is already checked in `select_utxos_for_consolidation` and so isn't required here.

Reply:
Also, I think there's a small bug below too:
```rust
select_utxos_from_pool(
						available_utxos,
						&bitcoin_fee_info,
						output_amount +
							number_of_outputs * fee_per_output_utxo +
							min_fee_required_per_tx,
						Some(Self::consolidation_parameters().consolidation_threshold),
```
It should also be consolidation_size here too? Here's what it says the meanings of the parameters are.

```rust
pub struct ConsolidationParameters {
	/// Consolidate when total UTXO count reaches this threshold
	pub consolidation_threshold: u32,
	/// Consolidate this many UTXOs
	pub consolidation_size: u32,
}
```

Reply:
Why not just do what I suggested and put all this inside the implementation of `select_utxos_for_consolidation`, and then inside there, first try the consolidation, and then if no consolidation is required, check for old utxos? It seems like we're handling a bunch of edge cases that we know will basically never happen. 

If we know we will do *either* 'consolidation above threshold' *or* 'consolidation of old utxos', then this is a lot easier to think about.

This way we would also make it easier to check whether the utxos are even worth rotating or not (net_value is currently not taken into account so we could rotate utxos that effectively unspendable).

Reply:
I think I didn't fully understand your design before. 
Now the logic is part of the consolidation, I also think this is cleaner. Please take a look and see if this is what you had in mind.

Reply:
> Also, I think there's a small bug below too:

WIth this, I was thinking about the case where we need more than consolidation_size number of utxos for the desired amount. Consider the edge case (which in practice should never happen)
e.g. threshold: 200, size: 100. If we use the "size" as the limit here:
we currently have 101 utxos, and we need all of them for the desired amount. We will need to wait until we have 99 more utxos before we consolidate - then we would have enough utxo.

I can see arguments for both - we should use "size" if we've decided that this is the maximum number of utxos we send period.
where as with threshold, it ensures that we can always find enough utxos for a transaction if we have enough. (though it is unlikely to happen)

what do you think? should we change from threshold to size?

Reply:
I had this discussion with @msgmaxim and from what I remember, the conclusion was that we should consider the consolidation threshold to be the largest number of utxos we are willing to sign (implying that we would use the threshold here, not the limit). In practice I don't think it makes a big difference, we just need to make a decision and stick to it.

Reply:
In this case I should leave the maximum as the threshold.

Reply:
Here is how I think about consolidation:
- If we need to make a transfer that will consume all of our UTXOs, we should be able to do it.
- For that, we ensure that the number of UTXO doesn't go above `N` (max we can sign at once), and if it does, we consolidate.
- We don't want to consolidate all at once (leaving not UTXO available for some time), so we consolidate `N' < N`.

In my consolidation PR I chose to call `N` - consolidation threshold and `N'` - consolidation size, and arbitrarily set `N' = N / 2`. We can change these if we want. This does mean that if we require to sign with `N`, we should allow that. If we find that `N` is too big for our engines, we should reduce *it* rather than limiting transactions to some number of UTXOs smaller than `N`.